Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1338 @@
+
+An example is `if true then 0 else 1 + 2`.
+Is this `if true then 0 else (1 + 2)` (shift) or `(if true then 0 else 1) + 2` (reduce)?
+
+You have a few choices here. The first is to split the expressions into even more productions, the second is to add a `%shift` directive to the `if-then-else`, and the third is to add an associativity and precedence to the `else`.
+
+I don't want to fiddle too much with the grammar or the LR parsing table, so for now, let's add a precedence to `else`:
+
+```diff
++%right else
+ %right '->'
+ %left '|'
+ %left '&'
+```
+
+This will cause the `else` keyword to extend as far to the right as possible, as it will have precedence over everything else.
+
+#### Parsing an operator atom
+
+This case is trivial, albeit annoying. Just shove these up in `atom`:
+
+```happy
+  -- Arithmetic operators
+  | '(' '+' ')'              { EOp (L.rtRange $1 <-> L.rtRange $2) (Plus (L.rtRange $2)) }
+  | '(' '-' ')'              { EOp (L.rtRange $1 <-> L.rtRange $2) (Minus (L.rtRange $2)) }
+  | '(' '*' ')'              { EOp (L.rtRange $1 <-> L.rtRange $2) (Times (L.rtRange $2)) }
+  | '(' '/' ')'              { EOp (L.rtRange $1 <-> L.rtRange $2) (Divide (L.rtRange $2)) }
+  -- Comparison operators
+  | '(' '=' ')'              { EOp (L.rtRange $1 <-> L.rtRange $2) (Eq (L.rtRange $2)) }
+  | '(' '<>' ')'             { EOp (L.rtRange $1 <-> L.rtRange $2) (Neq (L.rtRange $2)) }
+  | '(' '<' ')'              { EOp (L.rtRange $1 <-> L.rtRange $2) (Lt (L.rtRange $2)) }
+  | '(' '<=' ')'             { EOp (L.rtRange $1 <-> L.rtRange $2) (Le (L.rtRange $2)) }
+  | '(' '>' ')'              { EOp (L.rtRange $1 <-> L.rtRange $2) (Gt (L.rtRange $2)) }
+  | '(' '>=' ')'             { EOp (L.rtRange $1 <-> L.rtRange $2) (Ge (L.rtRange $2)) }
+  -- Logical operators
+  | '(' '&' ')'              { EOp (L.rtRange $1 <-> L.rtRange $2) (And (L.rtRange $2)) }
+  | '(' '|' ')'              { EOp (L.rtRange $1 <-> L.rtRange $2) (Or (L.rtRange $2)) }
+```
+
+#### Parsing let-in
+
+Finally, let's see `let...in...`, the final expression in our language.
+
+Let's add a new case to `exp` and see what happens:
+
+```happy
+  | dec in exp               { ELetIn (info $1 <-> info $3) $1 $3 }
+```
+
+12 shift/reduce conflicts. Not cool. However, looking at `Parser.info`, this is a case that we are already familiar with, which is how conditional expressions interacted with operators, although now with `let...in...` instead of `if...then...else...`.
+
+To illustrate, should `let a = b in a + b` be parsed as `(let a = b in a) + c` or `let a = b in (a + b)`? The second case is desirable, which is shifting when we reach `in`.
+
+The solution is simply to add a precedence for `in`:
+
+<!--
+TODO: I'm not entirely sure if there will be any problems in mixing else and in with the same precedence. Maybe one of the reviewers knows better.
```

Comment:
Not really. You just have `let ... in ... else ... let ... in ...` all associate to the right, as it should be: `let ... (in ... (else ... let ... (in ...)))`